### ISSUE TITLE
Training custom dataset

### ISSUE THREAD

**Issue by artuntun** [aripakman/amortized_community_detection#2]
Hi there,

Seems that the code has being customize to train on synthetically generated data where you can decide that each batch sample has the same number of datapoints per cluster. i.e the `K` is uniform within the batch. How can you adapt the model to train on less uniform data?

**Comment by yueqiw** [comment]
This is a great question!

The code fully supports mini-batches where batch samples have different numbers of clusters (K), different numbers of data points in each cluster, as well as different total numbers of data points (N). However, we have yet to release an example demonstrating this use case (we'll do that soon). 

You'll need to use `model.forward_train_mask(...)` rather than `model.forward_train(...)` as shown in the following lines:
https://github.com/aripakman/attentive_clustering_processes/blob/34ce5e5699be573098717f6bbc283733e062a0d9/acp/train_acp.py#L178

Here, the batch labels will be a list of length `batch_size` containing tensors of cluster labels for all batch samples. We use a `List[Tensor]` rather than 2D tensors because the number of data points in each sample may be different, but I think a 2D tensor of shape `[batch_size, num_points]` should also work. Note that you'll need to run `labels = [relabel(x) for x in labels]` before feeding the labels to the model, so that the first occurrence of each unique cluster ID is in ascending order.

Let us know if you have other questions.

**Comment by artuntun** [comment]
thank you for the fast answer :) That solve most of my doubts

If each of the samples have a variable number of datapoints how would you represent that on `data` tensor. The labels are `List[Tensor]` as you pointed out, but `data` is a `Tensor` with dimensions `[batch_size, num_points, feature_dimension]`. Do you need to pad with 0s the shorter samples?

**Comment by yueqiw** [comment]
For graphs, both DGL and torch_geometric has a batch data structure to represent a batch of graphs, so that they can be processed more efficiently by the GNN encoder, see below:
https://github.com/aripakman/attentive_clustering_processes/blob/34ce5e5699be573098717f6bbc283733e062a0d9/acp/data_generator/sbm_beta_generator.py#L109
The output of the GNN is then transformed into a list of tensors of variable shape:
https://github.com/aripakman/attentive_clustering_processes/blob/34ce5e5699be573098717f6bbc283733e062a0d9/acp/models/acp_model.py#L244

Then we copy these tensors to a fixed-size tensor of shape `[batch_size, max_num_points, feature_dimension]` with padding, and use a binary mask tensor to represent the non-empty indices. The neural network will use the mask tensor for downstream computations. 
https://github.com/aripakman/attentive_clustering_processes/blob/34ce5e5699be573098717f6bbc283733e062a0d9/acp/models/acp_model.py#L246-L253

If you're not working with graph data, you can concatenate the data tensors in the batch to shape  `[total_num_points, feature_dimension]`. The size of the labels will tell the model how to split the batch and create a mask tensor. Alternatively, a data tensor of shape `[batch_size, max_num_points, feature_dimension]` with padding should also work -- you'll need to skip the batch splitting operation `enc_data_raw = enc_data_raw.split(Ns, dim=0)`.

**Comment by artuntun** [comment]
Thanks, that worked!
